File: урок2/pointer.js
function counting(arr, k){
    let counter = 0;
    let last = 0;
    for(let first = 0; first < arr.length; first++){
        while(last < arr.length && arr[last] - arr[first] <= k){
            last++;
        }
        counter += arr.length - last;
    }
    return counter;
}

const readline = require('readline').createInterface({
    input: process.stdin,
    output: process.stdout
});

let input = [];
readline.on('line', (line) => {
    input.push(line);
    if (input.length === 2) {
        readline.close();
    }
}).on('close', () => {
    const [n, k] = input[0].split(' ').map(Number);
    const arr = input[1].split(' ').map(Number).slice(0, n);
    const result = counting(arr, k); 
    console.log(result); 
});

File: урок2/pointer.py
# дана отсортированная последовательность чисел N и число  K
# найти количество пар A, B, чтоб B - A > K

def cntrpairswithdiffgtk(sortednums, k):
    cntpairs = 0
    last = 0
    for first in range(len(sortednums)):
        while last < len(sortednums) and sortednums[last] - sortednums[first] <= k:
            last += 1
        cntpairs += len(sortednums) - last
    return cntpairs

print(cntrpairswithdiffgtk([1, 3, 11, 14, 17, 21, 34, 50], 10))
